kafka/adminapi.go

```diff
@@ -0,0 +1,964 @@
+	Type ResourceType
+	// Name of resource to set.
+	Name string
+	// Config entries to set.
+	// Configuration updates are atomic, any configuration property not provided
+	// here will be reverted (by the broker) to its default value.
+	// Use DescribeConfigs to retrieve the list of current configuration entry values.
+	Config []ConfigEntry
+}
+
+// String returns a human-readable representation of a ConfigResource
+func (c ConfigResource) String() string {
+	return fmt.Sprintf("Resource(%s, %s)", c.Type, c.Name)
+}
+
+// AlterOperation specifies the operation to perform on the ConfigEntry.
+// One of AlterOperationAdd, AlterOperationSet, AlterOperationDelete.
+type AlterOperation int
+
+const (
+	// AlterOperationAdd adds the configuration setting.
+	AlterOperationAdd = iota
+	// AlterOperationSet sets/overwrites the configuration setting.
+	AlterOperationSet = iota
+	// AlterOperationDelete deletes the configuration setting,
+	// possibly reverting it to default.
+	AlterOperationDelete = iota
+)
+
+// String returns the human-readable representation of an AlterOperation
+func (o AlterOperation) String() string {
+	switch o {
+	case AlterOperationAdd:
+		return "Add"
+	case AlterOperationSet:
+		return "Set"
+	case AlterOperationDelete:
+		return "Delete"
+	default:
+		return fmt.Sprintf("Unknown%d?", int(o))
+	}
+}
+
+// ConfigEntry holds parameters for altering a resource's configuration.
+type ConfigEntry struct {
+	// Name of configuration entry, e.g., topic configuration property name.
+	Name string
+	// Value of configuration entry.
+	Value string
+	// Operation to perform on the entry.
+	Operation AlterOperation
+}
+
+// StringMapToConfigEntries creates a new map of ConfigEntry objects from the
+// provided string map. The AlterOperation is set on each created entry.
+func StringMapToConfigEntries(stringMap map[string]string, operation AlterOperation) []ConfigEntry {
+	var ceList []ConfigEntry
+
+	for k, v := range stringMap {
+		ceList = append(ceList, ConfigEntry{Name: k, Value: v, Operation: operation})
+	}
+
+	return ceList
+}
+
+// String returns a human-readable representation of a ConfigEntry.
+func (c ConfigEntry) String() string {
+	return fmt.Sprintf("%v %s=\"%s\"", c.Operation, c.Name, c.Value)
+}
+
+// ConfigEntryResult contains the result of a single configuration entry from a
+// DescribeConfigs request.
+type ConfigEntryResult struct {
+	// Name of configuration entry, e.g., topic configuration property name.
+	Name string
+	// Value of configuration entry.
+	Value string
+	// Source indicates the configuration source.
+	Source ConfigSource
+	// IsReadOnly indicates whether the configuration entry can be altered.
+	IsReadOnly bool
+	// IsSensitive indicates whether the configuration entry contains sensitive information, in which case the value will be unset.
+	IsSensitive bool
+	// IsSynonym indicates whether the configuration entry is a synonym for another configuration property.
+	IsSynonym bool
```

Comment:
can this be derived by comparing the size of the `Synonyms` map to 1 (i.e. if a is a synonym of b then technically b is also a synonym of a) or is it the case that there are canonical names for which this property is never set to true (even if there is a synonym for the canonical name). If the former, I think this property should be removed. If the latter, i think this needs to be explicit in the property docs.

Reply:
maybe an integration test for this would be good.

Reply:
Synonyms is kind of a bad word for this, synonyms not only depicts actual synonyms (e.g., aliases) but also the same configuration property configured through other means. So when you set a property it will get a synonym pointing to the default-config property and value.

I'm not really sure what the value of these are from a user perspective, so I'm just going to stick with what the Java API does.